encoding/codecv3/codecv3.go

```diff
@@ -0,0 +1,249 @@
+	ParentBatchHash        common.Hash
+	LastBlockTimestamp     uint64
+	BlobDataProof          [64]byte
+	SkippedL1MessageBitmap []byte
+
+	// blob payload
+	blob *kzg4844.Blob
+	z    *kzg4844.Point
+}
+
+// NewDABatch creates a DABatch from the provided encoding.Batch.
+func NewDABatch(batch *encoding.Batch) (*DABatch, error) {
+	// this encoding can only support a fixed number of chunks per batch
+	if len(batch.Chunks) > codecv2.MaxNumChunks {
+		return nil, fmt.Errorf("too many chunks in batch")
+	}
+
+	if len(batch.Chunks) == 0 {
+		return nil, fmt.Errorf("too few chunks in batch")
+	}
+
+	if len(batch.Chunks[len(batch.Chunks)-1].Blocks) == 0 {
+		return nil, fmt.Errorf("too few blocks in last chunk of the batch")
+	}
+
+	// batch data hash
+	dataHash, err := codecv2.ComputeBatchDataHash(batch.Chunks, batch.TotalL1MessagePoppedBefore)
+	if err != nil {
+		return nil, err
+	}
+
+	// skipped L1 messages bitmap
+	bitmapBytes, totalL1MessagePoppedAfter, err := encoding.ConstructSkippedBitmap(batch.Index, batch.Chunks, batch.TotalL1MessagePoppedBefore)
+	if err != nil {
+		return nil, err
+	}
+
+	// blob payload
+	blob, blobVersionedHash, z, err := codecv2.ConstructBlobPayload(batch.Chunks, false /* no mock */)
+	if err != nil {
+		return nil, err
+	}
+
+	daBatch := DABatch{
+		Version:                uint8(encoding.CodecV3),
+		BatchIndex:             batch.Index,
+		L1MessagePopped:        totalL1MessagePoppedAfter - batch.TotalL1MessagePoppedBefore,
+		TotalL1MessagePopped:   totalL1MessagePoppedAfter,
+		DataHash:               dataHash,
+		BlobVersionedHash:      blobVersionedHash,
+		ParentBatchHash:        batch.ParentBatchHash,
+		LastBlockTimestamp:     batch.Chunks[len(batch.Chunks)-1].Blocks[len(batch.Chunks[len(batch.Chunks)-1].Blocks)-1].Header.Time,
+		SkippedL1MessageBitmap: bitmapBytes,
+		blob:                   blob,
+		z:                      z,
+	}
+
+	daBatch.BlobDataProof, err = daBatch.BlobDataProofForBatchHeader()
+	if err != nil {
+		return nil, err
+	}
+
+	return &daBatch, nil
+}
+
+// NewDABatchFromBytes attempts to decode the given byte slice into a DABatch.
+// Note: This function only populates the batch header, it leaves the blob-related fields empty.
+func NewDABatchFromBytes(data []byte) (*DABatch, error) {
+	if len(data) < 193 {
+		return nil, fmt.Errorf("insufficient data for DABatch, expected at least 193 bytes but got %d", len(data))
+	}
+
+	b := &DABatch{
+		Version:                data[0],
+		BatchIndex:             binary.BigEndian.Uint64(data[1:9]),
+		L1MessagePopped:        binary.BigEndian.Uint64(data[9:17]),
+		TotalL1MessagePopped:   binary.BigEndian.Uint64(data[17:25]),
+		DataHash:               common.BytesToHash(data[25:57]),
+		BlobVersionedHash:      common.BytesToHash(data[57:89]),
+		ParentBatchHash:        common.BytesToHash(data[89:121]),
+		LastBlockTimestamp:     binary.BigEndian.Uint64(data[121:129]),
+		SkippedL1MessageBitmap: data[193:],
+	}
+
+	copy(b.BlobDataProof[:], data[129:193])
```

Comment:
If we move this above line 124 then we'll get a type error?

Reply:
tried again and fixed in [4de8a88](https://github.com/scroll-tech/da-codec/pull/11/commits/4de8a88c6a939e5c0c8735346423e7ce4d2a2142).